### ISSUE TITLE
[SOLVED] Trying to understand the role of input_2 and lambda_1

### ISSUE THREAD

**Issue by LucasVandroux** [experiencor/basic-yolo-keras#165]
Hello @experiencor,

Sorry I have another question about the last two layers: `input_2` and `lambda_1`. To my understanding `reshape_1` is already outputting all the boxes with their dimensions, confidence and classes' scores. Therefore, why are `input_2` and `lambda_1` necessary? ([line 62-69](https://github.com/experiencor/basic-yolo-keras/blob/9b0ef52943a07d7b4e30007f1dfa93eae257c88b/frontend.py#L69) of `frontend.py`)
```python
# make the object detection layer
output = Conv2D(self.nb_box * (4 + 1 + self.nb_class), 
                              (1,1), strides=(1,1), 
                              padding='same', 
                              name='conv_23', 
                              kernel_initializer='lecun_normal')(features)
output = Reshape((self.grid_h, self.grid_w, self.nb_box, 4 + 1 + self.nb_class))(output)
output = Lambda(lambda args: args[0])([output, self.true_boxes])
```

Moreover, why is the last dimension of `lambda_1` equals to 4? ([line 39](https://github.com/experiencor/basic-yolo-keras/blob/9b0ef52943a07d7b4e30007f1dfa93eae257c88b/frontend.py#L39) of `frontend.py`)
```python
self.true_boxes = Input(shape=(1, 1, 1, max_box_per_image , 4))  
```

Is it linked to the computation of the loss function? If yes, how? Is it to limit the total number of predictions?

Again, thanks a lot for taking time to answer all of our questions 😸

**Comment by experiencor** [comment]
@LucasVandroux ```input_2``` holds ```self.true_boxes```, which is just a plain list of x,y,w,h of all the bounding boxes. This information is needed during custom lost computation and can actually be inferred from ```input_1``` (holding the ground truth) but it would take many tensorflow lines. So I use it to reduce the complexity of custom lost computation (the network becomes a little more complicated though).

```lambda_1``` is just a dummy layer used as a hack so that Keras does not complain about the unconnected network (```input_2``` does not feed to any layer of the network).

Yes. ```true_boxes``` contains x,y,w,h of all the bounding boxes in that image. So 4 is its last dimension.

Hope it clears your doubts.

**Comment by wyj2046** [comment]
Hi, @experiencor, I have a question in this line
`self.true_boxes = Input(shape=(1, 1, 1, max_box_per_image , 4))`
what every **1** means about? what axis they represent ?
Thanks!

**Comment by LucasVandroux** [comment]
@wyj2046 from @experiencor answer to my question I would say that the **1** are only here to respect whatever shape `input_2` needs to have for the custom loss computation hack to work.
